### ISSUE TITLE
Building mpv internally

### ISSUE THREAD

**Issue by saagarjha** [lhc70000/iina#1671]
(Continued from #1660)

So, my thoughts on a new build system would be to write up a shell script that just runs a compile of mpv and its dependencies. I'm considering creating an Xcode target for it as well, and possibly having a "update mpv" scheme that updates a precompiled build of mpv. It might be useful to fork mpv and add it as a submodule. What do you guys think?

**Comment by jerkstore369** [comment]
Also, as an observer, may I suggest.. maybe it's ok to fork mpv in that sooner or later mac apps are going to have to use metal without any choice in the matter, I think, right? If you could find someone willing... there was some talk about it on the mpv board but who knows if that'll come to fruition.

**Comment by saagarjha** [comment]
> sooner or later mac apps are going to have to use metal without any choice in the matter, I think, right

Do you have a source for this? I have never heard of anything to this effect.

**Comment by inflation** [comment]
I agree. Since we currently use brew's mpv builds, which are depends on specific machines. Having multiple targets built on a CI service could ease the compatibility issues.

**Comment by saagarjha** [comment]
Yes, I'm also interested in looking at this from a CI perspective. In fact, I had a plan in mind to get proper CI for IINA once we had a good build script landed ;)

I really want to discuss how we will implement this here. What are the goals we're looking to achieve? Here's my list, based on current experience, but feel free to comment on it:

For users:
1. We should have *one* IINA applications that users can download and run.
2. We should use the fasted possible implementation for each platform. Those with hardware acceleration should get it, and those with SSE and AVX should have it.
3. Ignoring codesigning, users sound be able to swap out components as they wish (though, we have no obligations to provide support!)

For developers:
1. Building lua should not be required for building a copy of IINA. That is, when I check out IINA from GitHub, I should have a binary already.
2. I *should* be able to swap this out myself. Ideally, it will build the source that I provide.
3. Building IINA should be nothing more than "click the run button in Xcode".

To meet performance requirements, I have a couple of ideas:

1. Have a bunch of different machines compile dependencies with `-march=native`. Ship with each of these, and load the correct one at runtime.
2. Gate features inside the library itself and have it perform the check itself. Load a single "psuedo-fat" library at runtime.
3. Compile options 1 and 2, creating different binaries for each architecture. Create a fat binary out of these binaries and load the single, resultant library.

To meet development goals, my solution is forking mpv and adding it as a submodule of IINA. This fork will contain a couple pre-built versions of mpv by default, which will be updated whenever we cut a new release. For day-to-day development, we will link directly against these. When testing new versions of mpv, or, when cutting a new release, someone will go to this submodule, fetch a newer commit or tag from upstream mpv, and build that. I'm considering adding a custom Xcode scheme that automates much of this process (not enabled by default, since building mpv will take too long for normal development, as far as I'm aware. Let me check and get back with my results.)

**Comment by saagarjha** [comment]
mpv took seven minutes to build on my computer. Could we consider having everyone who downloads IINA build their own copy?

**Comment by lhc70000** [comment]
> Ignoring codesigning, users sound be able to swap out components as they wish (though, we have no obligations to provide support!)

I think we could provide a brew formula based on the new build system, and advanced user can play with it in the ecosystem of Homebrew. For users who just want to double click IINA.app and play some videos, code signature 

> To meet performance requirements, I have a couple of ideas:

Since I have little experience on those stuff, I want to ask what we actually need to do to achieve this. Can we just supply some complier options, or we need to modify the build script or even the source code?

Overall this proposal looks great. How long will it take in your estimation? Is there anything I can help?

**Comment by saagarjha** [comment]
> I think we could provide a brew formula based on the new build system, and advanced user can play with it in the ecosystem of Homebrew

This isn't something I'm really keen on doing, because in my mind it's not really something that I feel users need. If advanced users want their own version of mpv, they can just use the one that Homebrew provides by default, no? This one will be optimized for their system anyways. Plus, there are a whole lot of other places to get these libraries: they could clone mpv and build them, or check out the IINA repository and build it from there.

Since you've given your approval for this, I'll get to work implementing it soon (probably not today or tomorrow, but hopefully by the end of the week. I still have to look at #1729 as well 😉). The eventual plan is to have a fork of mpv that we keep as a submodule, which I'll get hooked up through Xcode to build alongside the main IINA scheme. Stephen Canon was nice enough to answer [my question](https://stackoverflow.com/questions/45917280/mac-osx-minumum-support-sse-version#comment87738146_45921250) about the architecture slices we'll need to include, so I'll need to get this target to build a "normal" Intel as well as a Haswell slice in one fat binary. No changes to the source code should be necessary.

I'll push everything to a new branch forked off of develop. I really shouldn't need any help for this, and it should be done in a couple days–but if I don't get this done feel free to poke me to get this finished.

**Comment by lhc70000** [comment]
Any update?

**Comment by saagarjha** [comment]
I have a working shell script that builds mpv and is integrated with Xcode. I did have a question for you: what exactly do we need to build? I see that we need a custom FFmpeg as well, which itself depends on a lot of stuff. How many dependencies do we want to include?

**Comment by lhc70000** [comment]
That's why I gave up building mpv on my own and chose Homebrew……This is the complete dependency tree of mpv:

```
mpv
├── libass
│   ├── freetype
│   │   └── libpng
│   ├── fribidi
│   └── harfbuzz
│       ├── freetype
│       │   └── libpng
│       ├── graphite2
│       ├── icu4c
│       ├── cairo
│       │   ├── freetype
│       │   │   └── libpng
│       │   ├── fontconfig
│       │   │   └── freetype
│       │   │       └── libpng
│       │   ├── libpng
│       │   ├── pixman
│       │   └── glib
│       │       ├── gettext
│       │       ├── libffi
│       │       └── pcre
│       └── glib
│           ├── gettext
│           ├── libffi
│           └── pcre
├── ffmpeg
│   ├── lame
│   ├── x264
│   └── xvid
├── lua@5.1
├── jpeg
├── little-cms2
│   ├── jpeg
│   └── libtiff
│       └── jpeg
├── mujs
└── youtube-dl
```

We don't need a custom FFmpeg, but we need to build it from source to set `MACOSX_DEPLOYMENT_TARGET` to our desired version.

**Comment by saagarjha** [comment]
@lhc70000, how do you feel about this? Did you want me to work on this before 1.0 drops? Currently I think what we're looking at is an mpv submodule which we build with IINA, and compile with `-march=haswell` (for newer Macs) and `-march=core2 -mtune=sandybridge` (for older ones) and `lipo` the two slices together.
